src/multiselect_controls.js

```diff
@@ -0,0 +1,418 @@
+      const element = this.workspace_.getBlockById(id);
+      if (element) {
+        element.pathObject.updateSelected(true);
+      }
+    });
+  }
+  /**
+   * Enable the multiple select mode.
+   */
+  enableMultiselect() {
+    // Ensure that we only restore drag to move the workspace behavior
+    // when it is enabled.
+    if (!this.hasDisableWorkspaceDrag_ &&
+      this.workspace_.options.moveOptions &&
+      this.workspace_.options.moveOptions.drag) {
+      this.workspace_.options.moveOptions.drag = false;
+      this.hasDisableWorkspaceDrag_ = true;
+    }
+    this.dragSelect_ = new DragSelect({
+      selectables: document.querySelectorAll(
+          'g.blocklyDraggable:not(.blocklyInsertionMarker)' +
+        '> path.blocklyPath'),
+      area: document.querySelector('.blocklyWorkspace'),
+      multiselectMode: true,
+      draggability: false,
+      usePointerEvents: true,
+    });
+    this.dragSelect_.subscribe('elementselect', (info) => {
+      const element = info.item.parentElement;
+      if (inMultipleSelectionMode && element.dataset &&
+        element.dataset.id) {
+        this.updateBlocks_(
+            this.workspace_.getBlockById(element.dataset.id));
+      }
+    });
+    this.dragSelect_.subscribe('elementunselect', (info) => {
+      const element = info.item.parentElement;
+      if (inMultipleSelectionMode && element.dataset &&
+        element.dataset.id) {
+        this.updateBlocks_(
+            this.workspace_.getBlockById(element.dataset.id));
+      }
+    });
+    this.updateMultiselectIcon(true);
+    setSelectionMode(true);
+  }
+  /**
+   * Disable the multiple select mode.
+   */
+  disableMultiselect() {
+    setSelectionMode(false);
+    if (this.dragSelect_) {
+      this.dragSelect_.stop();
+      this.dragSelect_ = null;
+    }
+    // Ensure that at least Blockly select one of the blocks in the
+    // selection set, or clear the Blockly selection if our set is empty.
+    if (!blockSelection.size && Blockly.selected) {
+      Blockly.common.setSelected(null);
+    } else if (blockSelection.size && !Blockly.selected) {
+      Blockly.common.setSelected(
+          this.workspace_.getBlockById(blockSelection.keys().next().value));
+    }
+    if (this.hasDisableWorkspaceDrag_) {
+      this.workspace_.options.moveOptions.drag = true;
+      this.hasDisableWorkspaceDrag_ = false;
+    }
+    this.updateMultiselectIcon(false);
+  }
+  /**
+   * Updates the multi select icon.
+   * @param {boolean} enable Whether the multi select is enabled.
+   */
+  updateMultiselectIcon(enable) {
+    if (!this.multiselectGroup_) {
+      return;
+    }
+    this.enabled = enable;
+    if (enable) {
+      this.multiselectGroup_.firstElementChild.setAttributeNS(
+          Blockly.utils.dom.XLINK_NS, 'xlink:href', ENABLED_IMG);
+    } else {
+      this.multiselectGroup_.firstElementChild.setAttributeNS(
+          Blockly.utils.dom.XLINK_NS, 'xlink:href', DISABLED_IMG);
+    }
```

Comment:
Can there potentially be 3 states? Multiselect on (gestures will add/remove blocks to the selection) Multiselect off (gestures trigger drags, etc.) and disabled (cannot turn on multiselect with control or shift key). Maybe this something to consider for a later version but if we want to allow it agreeing on terminology (on/off/disabled vs enabled/disabled) would be good.

Reply:
I guess for disabled in on/off/disabled, as it's only a plugin and not some upstream features, we can call `plugin.dispose()` to disable the plugin, maybe there's no need for disabled in on/off/disabled, only on/off is enough.